solr/core/src/java/org/apache/solr/cloud/ZkController.java

```diff
@@ -1703,7 +1703,8 @@ public void publish(final CoreDescriptor cd, final Replica.State state, boolean
       DocCollection coll = zkStateReader.getCollection(collection);
       if (forcePublish || sendToOverseer(coll, coreNodeName)) {
         overseerJobQueue.offer(Utils.toJSON(m));
-      } else {
+      }
+      if (coll != null && coll.isPerReplicaState() && coreNodeName != null) {
```

Comment:
I feel we should wait here to update state.json file?

Reply:
@noblepaul / @chatman any thought on this!

Reply:
Why? Why should it wait?

Reply:
1. nonprs collection updates state.json atomically.
2. With prs here, we will update synchronously, but there is possibility that state.json will be updated later.

Reply:
Even if it's not refreshed atomically, it should be just fine

Reply:
In PRS, we update structure of collection first in stats.json, and then update the prs state. We have written go-client with such assumption. I don't think we want to break this assumption of prs protocol.

Reply:
The data that's updated by the data node (prs states) is not the same as the data updated by overseer (state.json). Hence, they can be refreshed independently. Also, we should never wait for overseer to execute the state messages. It's horrible for performance
We should write our go client consistent with how Solr and SolrJ behaves and not the other way around

Reply:
Possibly, data updated in this case is different. But we should keep behavior similar to what non-prs have in this case. Otherwise we might run in some race-conditions, where code is waiting for updates of (shard and replica).

Please let me know for what we have changed this piece of code now?

Reply:
This doesn't matter for us. It's only relevant for the JUnit tests. So, the messages sent to overseer becomes no-op

Reply:
Are you saying we will never send message to overseer for prs-collection?

I do see during split, it sends message to overseer and then it updates prs states.

Reply:
we send messages to overseer in certain cases. but, they don't result in updating PRS states

Reply:
We have removed the else part that means there is case where we send two msg one by one?

Also, I observed that during split msg is going to overseer and we are also updating prs state. That's why I'm saying we need to wait here.

Would  be good to write unit test to verify this.

Reply:
> We have removed the else part that means there is case where we send two msg one by one?

We send one mesage to overseer and do one direct update of PRS states. The overseer **may** update some attributes of the replica as a result of the message but it never touches the PRS states

> Also, I observed that during split msg is going to overseer and we are also updating prs state. That's why I'm saying we need to wait here.

We should ensure that overseer never updates PRS states . 

> Would be good to write unit test to verify this.

yes. let's add testcases to ensure this